Source/Core/Core/IOS/USB/Emulated/Microphone.cpp

```diff
@@ -0,0 +1,261 @@
+// Copyright 2024 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/IOS/USB/Emulated/Microphone.h"
+
+#include <algorithm>
+
+#include <cubeb/cubeb.h>
+
+#include "AudioCommon/CubebUtils.h"
+#include "Common/Event.h"
+#include "Common/Logging/Log.h"
+#include "Common/ScopeGuard.h"
+#include "Common/Swap.h"
+#include "Core/Config/MainSettings.h"
+#include "Core/Core.h"
+#include "Core/IOS/USB/Emulated/WiiSpeak.h"
+
+#ifdef _WIN32
+#include <Objbase.h>
+#endif
+
+#ifdef ANDROID
+#include "jni/AndroidCommon/IDCache.h"
+#endif
+
+namespace IOS::HLE::USB
+{
+Microphone::Microphone(const WiiSpeakState& sampler) : m_sampler(sampler)
+{
+#ifdef _WIN32
+  Common::Event sync_event;
+  m_work_queue.EmplaceItem([this, &sync_event] {
+    Common::ScopeGuard sync_event_guard([&sync_event] { sync_event.Set(); });
+    auto result = ::CoInitializeEx(nullptr, COINIT_MULTITHREADED | COINIT_DISABLE_OLE1DDE);
+    m_coinit_success = result == S_OK;
+    m_should_couninit = result == S_OK || result == S_FALSE;
+  });
+  sync_event.Wait();
+#endif
+
+  StreamInit();
+}
+
+Microphone::~Microphone()
+{
+  StreamTerminate();
+
+#ifdef _WIN32
+  if (m_should_couninit)
+  {
+    Common::Event sync_event;
+    m_work_queue.EmplaceItem([this, &sync_event] {
+      Common::ScopeGuard sync_event_guard([&sync_event] { sync_event.Set(); });
+      m_should_couninit = false;
+      CoUninitialize();
+    });
+    sync_event.Wait();
+  }
+  m_coinit_success = false;
+#endif
+}
+
+void Microphone::StreamInit()
+{
+#ifdef _WIN32
+  if (!m_coinit_success)
+  {
+    ERROR_LOG_FMT(IOS_USB, "Failed to init Wii Speak stream");
+    return;
+  }
+  Common::Event sync_event;
+  m_work_queue.EmplaceItem([this, &sync_event] {
+    Common::ScopeGuard sync_event_guard([&sync_event] { sync_event.Set(); });
+#endif
+    m_cubeb_ctx = CubebUtils::GetContext();
+#ifdef _WIN32
+  });
+  sync_event.Wait();
+#endif
```

Comment:
This pattern occurs several times in this file. I recommend creating a helper function that handles the boilerplate and lets you pass in a lambda.